lsp-types/src/Language/LSP/MetaModel/CodeGen.hs

```diff
@@ -0,0 +1,911 @@
+-}
+
+-- | A mapping from names in the metamodel to their names in the generated Haskell.
+type SymbolTable = Map.Map Text TH.Name
+
+-- | A mapping from names in the metamodel to their structure definition, used for chasing
+-- supertypes.
+type StructTable = Map.Map Text Structure
+
+data CodeGenEnv = CodeGenEnv {
+  symbolTable :: SymbolTable
+  , structTable :: StructTable
+  }
+type CodeGenM = ReaderT CodeGenEnv TH.Q
+
+-- | Given the metamodel, make a splice with declarations for all the types.
+genMetaModel :: MetaModel -> TH.Q [ TH.Dec ]
+genMetaModel mm = do
+  (symbolTable, structTable) <- buildTables mm
+  flip runReaderT (CodeGenEnv symbolTable structTable) $ do
+    structs <- concat <$> traverse genStruct (structures mm)
+    -- Make a binding that just lists all the struct names so we can use them later
+    -- See Note [Generating lenses]
+    structNames <- do
+      -- Have to use the string form of the generated Name
+      -- since we might have mangled the original name
+      let structNames = mapMaybe (\Structure{name} -> Map.lookup name symbolTable) (structures mm)
+      structNameEs <- lift $ traverse (TH.lift . TH.nameBase) structNames
+      lift
+        [d|
+          structNames :: [String]
+          structNames = $(pure $ TH.ListE structNameEs)
+        |]
+    enums <- concat <$> traverse genEnumeration (enumerations mm)
+    aliases <- traverse genAlias (typeAliases mm)
+    methods <- genMethods (requests mm) (notifications mm)
+    pure (structs ++ structNames ++ enums ++ aliases ++ methods)
+
+  -- | Convenience wrapper around 'genMetaModel' that reads the input from a file.
+genMetaModelFromFile :: FilePath -> TH.Q [ TH.Dec ]
+genMetaModelFromFile fp = do
+  TH.addDependentFile fp
+  res <- liftIO $ J.eitherDecodeFileStrict' fp
+  case res of
+    Left e   -> fail e
+    Right mm -> genMetaModel mm
+
+-- | Given a list of type names, make a splice that generates the lens typeclass declarations
+-- for all of them. Defined here to avoid stage restrictions.
+genLenses :: [String] -> TH.Q [TH.Dec]
+genLenses structNames = do
+  let structNames' = fmap TH.mkName structNames
+  let
+    -- We need to use the internals of the lens TH machinery so that we can do this
+    -- in one go without generating duplicate classes.
+    opticMaker :: TH.Name -> HasFieldClasses [TH.Dec]
+    opticMaker n = do
+        (TH.TyConI d) <- lift $ TH.reify n
+        makeFieldOpticsForDec' classUnderscoreNoPrefixFields d
+  decss <- flip evalStateT mempty $ traverse opticMaker structNames'
+  pure $ concat decss
+
+------------
+
+-- | Names we can't put in Haskell code.
+reservedNames :: Set.Set T.Text
+reservedNames = Set.fromList [ "data", "type" ]
+
+-- | Sanitize a name so we can use it in Haskell.
+sanitizeName :: T.Text -> T.Text
+sanitizeName n =
+  -- Names can't start with underscores! Replace that with a 'U' for lack
+  -- of a better idea
+  let n' = if "_" `T.isPrefixOf` n then T.cons 'U' $ T.tail n else n
```

Comment:
What about `Abstract` or `Base` or `Internal` or something like that. Looks like most usages are "abstract" classes in the traditional OOP sense.